app/Http/Controllers/Manager/WalletController.php

```diff
@@ -328,6 +329,108 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
         return $this->withdrawAds($request, $rpcClient);
     }
 
+    private function getSwashBSCAddress(): WalletAddress
+    {
+        try {
+            return new WalletAddress('BSC', config('app.swash_bsc_address'));
+        } catch (InvalidArgumentException $e) {
+            Log::error(sprintf('Invalid BSC address is set: %s', $e->getMessage()));
+            throw new HttpException(Response::HTTP_INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    public function withdrawBatchSwash(Request $request, AdsRpcClient $rpcClient): JsonResponse
+    {
+        /** @var User $user */
+        $user = Auth::user();
+
+        if(!$user->isModerator()){
+            return self::json(['Auth' => 'Only Moderator/Admin can run this API'], JsonResponse::HTTP_FORBIDDEN);
+        }
+
+        $addressFrom = $this->getAdServerAdsAddress();
+        $address = $this->getSwashBSCAddress();
+        $addressTo = $this->getWalletAdsAddress($rpcClient, $address);
+        $message = $this->getWalletAdsMessage($rpcClient, $address);
+        $batchId = hash('sha256', strval(microtime(true)) );
+
+        $balanceTotal = UserLedgerEntry::getWalletBalanceForAllUsers();
+        $users_balances = UserLedgerEntry::allWalletBalanceIfAny();
+
+        $amount = AdsUtils::calculateAmount($addressFrom, $addressTo, $balanceTotal);
+        $adsFee = AdsUtils::calculateFee($addressFrom, $addressTo, $amount);
+        $total = $amount + $adsFee;
+
+        if ($balanceTotal < $total) {
+            throw new UnprocessableEntityHttpException(sprintf('Insufficient total: %d, fee: %d', $balanceTotal, $adsFee));
+        }
+        DB::beginTransaction();
+        
+        foreach ($users_balances as $item) {
+            $user_item = User::fetchById($item['uid']);
+            $ledgerEntry = UserLedgerEntry::constructSwash(
+                $batchId,
+                $user_item->id,
+                -$item['share']
+            )->addressed($addressFrom, $addressTo);
+            if (!$ledgerEntry->save()) {
+                DB::rollBack();
+                throw new InternalErrorException();
+            }
+        }
+        AdsSendOneSwash::dispatch(
+            $batchId,
+            $addressTo,
+            $amount,
+            $message
+        );
+        DB::commit();
```

Comment:
What will happen if previous command works and send users money with executing the transaction but it fails to commit DB changes? Will you pay the users again in the next period?

Reply:
The code is written based on `withdrawAds` line 504.
However, I swapped the commit and dispatch. I also added a log message to keep a track of the user shares.